File: trunk/simplehsm_c/oven.c
#include <stdio.h>
#include <tchar.h>

#include "simplehsm.h"

//
// Oven State machine signals
//

enum
{
  SIG_OPEN_DOOR = SIG_USER,
} OvenSignals;

//
// Oven State function definitions
//

stnext oven(int signal, void* param);
  stnext heating(int signal, void* param);
    stnext toasting(int signal, void* param);
    stnext baking(int signal, void* param);
  stnext doorOpen(int signal, void* param);

//
// Oven State function implementations
//

stnext oven(int signal, void* param)
{
  switch (signal)
  {
    case SIG_ENTRY:
      printf("oven: entering state\n");
      return stnone;
    case SIG_INIT:
      InitTransitionState(heating);
      return stnone;
    case SIG_EXIT:
      printf("oven: exiting state\n");
      return stnone;
  }
  return stnone;
}

stnext heating(int signal, void* param)
{
  switch (signal)
  {
    case SIG_ENTRY:
      printf("  heating: entering state\n");
      return stnone;
    case SIG_INIT:
      InitTransitionState(toasting);
      return stnone;
    case SIG_EXIT:
      printf("  heating: exiting state\n");
      return stnone;
    case SIG_OPEN_DOOR:
      printf("  heating: OPEN_DOOR signal!\n");
      TransitionState(doorOpen);
      return stnone;
  }
  return (stnext)oven;
}

stnext toasting(int signal, void* param)
{
  switch (signal)
  {
    case SIG_ENTRY:
      printf("    toasting: entering state\n");
      return stnone;
    case SIG_INIT:
      return stnone;
    case SIG_EXIT:
      printf("    toasting: exiting state\n");
      return stnone;
  }
  return (stnext)heating;
}

stnext baking(int signal, void* param)
{
  switch (signal)
  {
    case SIG_ENTRY:
      printf("    baking: entering state\n");
      return stnone;
    case SIG_INIT:
      return stnone;
    case SIG_EXIT:
      printf("    baking: exiting state\n");
      return stnone;
  }
  return (stnext)heating;
}

stnext doorOpen(int signal, void* param)
{
  switch (signal)
  {
    case SIG_ENTRY:
      printf("  doorOpen: entering state\n");
      return stnone;
    case SIG_INIT:
      return stnone;
    case SIG_EXIT:
      printf("  doorOpen: exiting state\n");
      return stnone;
  }
  return (stnext)oven;
}

void ShowStatus(void)
{
  printf("\nChecking States:\n");
  printf("  in oven state: %d\n  in heating state: %d\n  in toasting state: %d\n  in baking state: %d\n  in doorOpen state: %d\n\n",
         IsInState(oven), IsInState(heating), IsInState(toasting), IsInState(baking), IsInState(doorOpen));
}

int _tmain(int argc, _TCHAR* argv[])
{
  InitialState(oven);
  SignalCurrentState(SIG_OPEN_DOOR, NULL);
  ShowStatus();
	return 0;
}

File: trunk/simplehsm_python/oven.py
import simplehsm

#
# Oven State machine signals
#

SIG_OPEN_DOOR = simplehsm.SIG_USER,

#
# Oven State hierachy
#

#  oven
#    heating
#      toasting
#      baking
#    doorOpen

#
# Oven state machine
#

class Oven(simplehsm.SimpleHsm):

    def __init__(self):
        self.InitialState(self.oven);

    #
    # Oven State implementations
    #

    def oven(self, state_event):
        if (state_event.sig == simplehsm.SIG_ENTRY):
            print "oven: entering state"
            return None
        elif (state_event.sig == simplehsm.SIG_INIT):
            self.InitTransitionState(self.heating)
            return None
        elif (state_event.sig == simplehsm.SIG_EXIT):
            print "oven: exiting state"
            return None
        return None

    def heating(self, state_event):
        global SIG_OPEN_DOOR
        if (state_event.sig == simplehsm.SIG_ENTRY):
            print "  heating: entering state"
            return None
        elif (state_event.sig == simplehsm.SIG_INIT):
            self.InitTransitionState(self.toasting);
            return None
        elif (state_event.sig == simplehsm.SIG_EXIT):
            print "  heating: exiting state"
            return None
        elif (state_event.sig == SIG_OPEN_DOOR):
            print "  heating: OPEN_DOOR signal!"
            self.TransitionState(self.doorOpen)
            return None;
        return self.oven;

    def toasting(self, state_event):
        if (state_event.sig == simplehsm.SIG_ENTRY):
            print "    toasting: entering state"
            return None
        elif (state_event.sig == simplehsm.SIG_INIT):
            return None
        elif (state_event.sig == simplehsm.SIG_EXIT):
            print "    toasting: exiting state"
            return None
        return self.heating

    def baking(self, state_event):
        if (state_event.sig == simplehsm.SIG_ENTRY):
            print "    baking: entering state"
            return None
        elif (state_event.sig == simplehsm.SIG_INIT):
            return None
        elif (state_event.sig == simplehsm.SIG_EXIT):
            print "    baking: exiting state"
            return None
        return self.heating

    def doorOpen(self, state_event):
        if (state_event.sig == simplehsm.SIG_ENTRY):
            print "  doorOpen: entering state"
            return None;
        elif (state_event.sig == simplehsm.SIG_INIT):
            return None;
        elif (state_event.sig == simplehsm.SIG_EXIT):
            print "  doorOpen: exiting state"
            return None;
        return self.oven

    def ShowStatus(self):
        print "\nChecking States:"
        print "  in oven state: %d\n  in heating state: %d\n  in toasting state: %d\n  in baking state: %d\n  in doorOpen state: %d\n\n" % (self.IsInState(self.oven), self.IsInState(self.heating), self.IsInState(self.toasting), self.IsInState(self.baking), self.IsInState(self.doorOpen))

def main():
    oven = Oven()
    oven.SignalCurrentState(simplehsm.StateEvent(SIG_OPEN_DOOR))
    oven.ShowStatus()

main()
